### PR TITLE
projects: max14906 : Add support for STM32

### PR DESCRIPTION
**Pull request by RaduSabau1**
## Pull Request Description

This PR is adding STM32 platform support for the MAX14906 application example.
Other application example files than the stm32 platform related files have been modified in order to maintain versatility.

## PR Type
- [ ] Bug fix (change that fixes an issue)
- [x] New feature (change that adds new functionality)
- [ ] Breaking change (has dependencies in other repos or will cause CI to fail)

## PR Checklist
- [x] I have followed the [Coding style guidelines](http://analogdevicesinc.github.io/no-OS/drivers_guide.html#coding-style)
- [ ] I have performed a self-review of the changes
- [x] I have commented my code, at least hard-to-understand parts
- [x] I have build all projects affected by the changes in this PR
- [x] I have tested in hardware affected projects, at the relevant boards
- [x] I have signed off all commits from this PR
- [ ] I have updated the documentation (wiki pages, ReadMe etc), if applies

### CODE REVIEW

File: projects/max14906/src/platform/maxim/main.c [hunk 1]
```diff
@@ -58,11 +58,30 @@ int main()
 
 #ifdef BASIC_EXAMPLE
 	struct no_os_uart_desc *uart_desc;
+	struct no_os_irq_ctrl_desc *gpio_irq_desc;
```
Comment (buha) [inline projects/max14906/src/platform/maxim/main.c:61]: I think all the gpio irq related code that you put in this maxim/main.c file should go to basic_example.c


And in parameters.h/.c (bsp) you should provide GPIO_IRQ_ID, GPIO_IRQ_OPS and some common macro for GPIO0_IRQn and it's equivalent for stm32.

Comment (rbolboac) [reply projects/max14906/src/platform/maxim/main.c:61]: There are discrepancies between maxim and stm when treating interrupts and I think these changes currently are related only to maxim platform

File: projects/max14906/Makefile [hunk 2]
```diff
@@ -1,6 +1,6 @@
 # Select the example you want to enable by choosing y for enabling and n for disabling
-BASIC_EXAMPLE = n
-IIO_EXAMPLE = y
+BASIC_EXAMPLE = y
```
Comment (amiclaus) [inline projects/max14906/Makefile:2]: unrelated change.

File: projects/max14906/src/common/common_data.h [hunk 3]
```diff
@@ -49,7 +49,10 @@
 
 extern struct no_os_uart_init_param max14906_uart_ip;
 extern struct no_os_spi_init_param max14906_spi_ip;
-extern struct no_os_gpio_init_param max14906_fault_gpio_param;
 extern struct max149x6_init_param max14906_ip;
 
+#ifdef BASIC_EXAMPLE
+extern struct no_os_gpio_init_param max14906_fault_gpio_param;
```
Comment (amiclaus) [inline projects/max14906/src/common/common_data.h:55]: is this change related to adding stm32 support?

File: projects/max14906/src/platform/maxim/parameters.h [hunk 4]
```diff
@@ -42,6 +42,7 @@
 #include "maxim_irq.h"
 #include "maxim_spi.h"
 #include "maxim_gpio.h"
+#include "maxim_gpio_irq.h"
```
Comment (amiclaus) [inline projects/max14906/src/platform/maxim/parameters.h:45]: is this related to adding stm32 support?

File: projects/max14906/src/platform/maxim/parameters.h [hunk 5]
```diff
@@ -80,6 +81,7 @@ extern struct max_spi_init_param max14906_spi_extra;
 #define GPIO_OPS	&max_gpio_ops
 #define GPIO_EXTRA	&max14906_gpio_extra_ip
 #define GPIO_IRQ_OPS	&max_gpio_irq_ops
+#define GPIO_IRQ_EXTRA	NULL
```
Comment (amiclaus) [inline projects/max14906/src/platform/maxim/parameters.h:84]: ditto.

File: projects/max14906/src/platform/maxim/parameters.h (hunk 4 above)
Comment (RaduSabau1) [reply projects/max14906/src/platform/maxim/parameters.h:45]: yes, because the new basic_example.c after adding support for STM platform requires in the global_desc_param the
GPIO_IRQ_OPS and GPIO_IRQ_EXTRA macros.
GPIO_IRQ_OPS for maxim is max_gpio_irq_ops which is to be found in maxim_gpio_irq.h
GPIO_IRQ_EXTRA in case of maxim should be null.

File: projects/max14906/src/platform/maxim/parameters.h (hunk 5 above)
Comment (RaduSabau1) [reply projects/max14906/src/platform/maxim/parameters.h:84]: GPIO_IRQ_EXTRA in case of maxim should be null.

File: projects/max14906/src/common/common_data.h (hunk 3 above)
Comment (RaduSabau1) [reply projects/max14906/src/common/common_data.h:55]: Not really, will revert the change.

File: projects/max14906/Makefile (hunk 2 above)
Comment (RaduSabau1) [reply projects/max14906/Makefile:2]: Reverting it.

**Comment by RaduSabau1** [comment]
V2

- Removed unrelated changes.

**Comment by RaduSabau1** [comment]
V3

- Rebased with master.

File: projects/max14906/src/platform/maxim/main.c (hunk 1 above)
Comment (buha) [reply projects/max14906/src/platform/maxim/main.c:61]: ok, in that case we keep them as they are but we need to fix the discrepancies separately and i would say ASAP because we keep adding projects that we will need to refactor...

**Comment by RaduSabau1** [comment]
V4

- Rebased with master.

File: projects/max14906/src/examples/basic/basic_example.c [hunk 6]
```diff
@@ -164,7 +146,7 @@ int basic_example_main()
 	/** Read current limit for all channels. */
 	for (i = 0; i < MAX14906_CHANNELS; i++) {
 		ret = max14906_climit_get(max14906_desc, i, &climit2);
-		if(ret)
+		if (ret)
```
Comment (rbolboac) [inline projects/max14906/src/examples/basic/basic_example.c:149]: unrelated changes, please create a new commit for this

File: projects/max14906/src/examples/basic/basic_example.c [hunk 7]
```diff
@@ -173,16 +155,18 @@ int basic_example_main()
 	/** Setting the on state for channel 0, and then verifying it. */
 	if (max14906_desc->en_gpio) {
 		ret = max14906_ch_set(max14906_desc, 0, 1);
-		if(ret)
+		if (ret)
```
Comment (rbolboac) [inline projects/max14906/src/examples/basic/basic_example.c:158]: unrelated changes, please create a new commit for this

File: projects/max14906/src/platform/stm32/main.c [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+	struct no_os_uart_desc *uart;
+
+	ret = no_os_uart_init(&uart, &max14906_uart_ip);
+	if (ret)
+		goto error;
+
+	no_os_uart_stdio(uart);
+	ret = basic_example_main();
+	if (ret)
+		goto error;
```
Comment (rbolboac) [inline projects/max14906/src/platform/stm32/main.c:79]: uart_remove should be called and can remove these lines (keep the same logic as for the other examples)

File: projects/max14906/src/platform/maxim/main.c [hunk 9]
```diff
@@ -58,11 +58,30 @@ int main()
...
+		return ret;
+	}
+
+	ret = no_os_irq_enable(gpio_irq_desc, GPIO0_IRQn);
+	if (ret) {
+		no_os_irq_ctrl_remove(gpio_irq_desc);
+		no_os_uart_remove(uart_desc);
+		return ret;
+	}
+
```
Comment (rbolboac) [inline projects/max14906/src/platform/maxim/main.c:84]: let's use the same logic with the labels

**Review by rbolboac — COMMENTED** [review]
some comments from my side...
also, I would split these changes in two different commit.
first I would add the interrupt support for maxim and only after I would add support for stm32 platform
then there are also some style related changes which should be performed in a different commit

File: projects/max14906/src/examples/basic/basic_example.c (hunk 7 above)
Comment (RaduSabau1) [reply projects/max14906/src/examples/basic/basic_example.c:158]: Adding new commit.

File: projects/max14906/src/platform/stm32/main.c [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+ * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+ * IN NO EVENT SHALL ANALOG DEVICES BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, INTELLECTUAL PROPERTY RIGHTS, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*******************************************************************************/
+#include <errno.h>
```
Comment (rbolboac) [inline projects/max14906/src/platform/stm32/main.c:39]: is this include needed?

File: projects/max14906/src/platform/stm32/main.c [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+{
+	int ret = -ENODEV;
+
+	max14906_spi_extra_ip.get_input_clock = HAL_RCC_GetPCLK1Freq;
+	stm32_init();
+
+#ifdef IIO_EXAMPLE
+	ret = iio_example_main();
+	if (ret)
+		goto error;
```
Comment (rbolboac) [inline projects/max14906/src/platform/stm32/main.c:66]: I think we can remove these lines

**Review by rbolboac — APPROVED** [review]
just too small remarks, all in all looks good

File: projects/max14906/src/platform/stm32/main.c (hunk 10 above)
Comment (RaduSabau1) [reply projects/max14906/src/platform/stm32/main.c:39]: no, will remove it.

File: projects/max14906/src/platform/stm32/main.c (hunk 11 above)
Comment (RaduSabau1) [reply projects/max14906/src/platform/stm32/main.c]: Removing them.

File: projects/max14906/src/platform/stm32/main.c [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+#endif
+
+/*******************************************************************************
+ * @brief Main function execution for STM32 platform.
+ *
+ * @return ret - Result of the enabled examples execution.
+*******************************************************************************/
+int main()
+{
+	int ret = -ENODEV;
```
Comment (rbolboac) [inline projects/max14906/src/platform/stm32/main.c:57]: so it looks like the build is failing after removing errno.h because of this line. I would however use #include "no_os_error.h", similarly to the main fail in maxim platform, and also, initialize the ret value with -EINVAL.

**Comment by RaduSabau1** [comment]
V6

- Fixed main source file errors for stm platform.

File: projects/max14906/src/platform/stm32/main.c (hunk 12 above)
Comment (RaduSabau1) [reply projects/max14906/src/platform/stm32/main.c]: Sure.

**Review by amiclaus — APPROVED** [review]

**Review by rbolboac — APPROVED** [review]
